### PR TITLE
ADDED: proper test case for RelativePositionalEncodingLayer

### PR DESCRIPTION
**Pull request by npwynands**
Related to issue #327

### CODE REVIEW

File: returnn/tf/layers/rec.py [hunk 1]
```diff
@@ -8123,6 +8123,7 @@ def get_out_data_from_opts(cls, name, sources, n_out, **kwargs):
     data = data.copy_template().copy_as_batch_major()
     data.batch_dim_axis = None
     data.feature_dim_axis = NotSpecified
+    data.dim = n_out
```
Comment (albertz) [inline returnn/tf/layers/rec.py:8126]: This was already committed, right?
Please rebase on current master.

**Comment by albertz** [comment]
You reformatted the whole file. Don't do that. Please just add your test, without changing anything else. (So delete all the existing commits, and then do a new commit which just adds the test.)

**Comment by albertz** [comment]
Why are there multiple commits? This is just a single test case. Just use a single commit.

**Comment by albertz** [comment]
Your formatting is wrong. We use two space indents. Please always stay consistent to existing code! See the contribution guidelines. Also see PyCharm settings in wiki.

**Comment by npwynands** [comment]
Which commit message is too long?

**Comment by albertz** [comment]
> Which commit message is too long?

You see that here on GitHub. Just click on commits. It will show you which is too long by removing the trail and adding "...".

**Comment by npwynands** [comment]
Ah, you mean the "REMOVED: [...]"? Okay.

**Comment by npwynands** [comment]
Okay, I changed the space indent and squashed my commits. Is it fine now?

**Comment by albertz** [comment]
The commit message is still somewhat inconsistent to other commit messages. We never have sth like "ADDED" as a prefix. Please keep it consistent. In this case, I would just call it "test_RelativePositionalEncodingLayer".

**Comment by albertz** [comment]
Otherwise looks fine now. So this test failed before and runs fine now?

**Comment by JackTemaki** [comment]
> Ah, you mean the "REMOVED: [...]"? Okay.

If you write your commit messages with git commit using vim, it will show you the allowed commit message length in yellow, and exceeding characters in white.

**Comment by albertz** [comment]
> > Ah, you mean the "REMOVED: [...]"? Okay.
> 
> If you write your commit messages with git commit using vim, it will show you the allowed commit message length in yellow, and exceeding characters in white.

With PyCharm as well, when you configure it correctly.

**Comment by JackTemaki** [comment]
> With PyCharm as well, when you configure it correctly.

I did not know that the configuration option existed. I will add this as well. Is 72 the correct maximum?

**Comment by albertz** [comment]
> > With PyCharm as well, when you configure it correctly.
> 
> I did not know that the configuration option existed.

Well, this was kind of obvious that it does, or not?
I just was a bit surprised that it is not enabled by default. Although I think it was in some earlier version.

> I will add this as well.

I already added it.

> Is 72 the correct maximum?

I don't really know. I think so.

**Comment by npwynands** [comment]
> The commit message is still somewhat inconsistent to other commit messages. We never have sth like "ADDED" as a prefix. Please keep it consistent. In this case, I would just call it "test_RelativePositionalEncodingLayer".

I don’t see any pattern/convention in your commit messages and you already merged one of my commits where I’ve been using such a prefix (8ce5f9494d9b379bc4713aa4c5387f13846421cf). I will remember that for future commits, but I don’t think that matters here...

**Comment by JackTemaki** [comment]
> > Is 72 the correct maximum?
> 
> I don't really know. I think so.

It is 50 for the subject and 72 for the body, I changed it accordingly

**Comment by npwynands** [comment]
> > Ah, you mean the "REMOVED: [...]"? Okay.
> 
> If you write your commit messages with git commit using vim, it will show you the allowed commit message length in yellow, and exceeding characters in white.

Ah okay, I simply write my commit messages into the terminal. There are no such assisting tools.

**Comment by npwynands** [comment]
> Otherwise looks fine now. So this test failed before and runs fine now?

Yes, it does fail due to the sanity check.

**Comment by albertz** [comment]
> > > Is 72 the correct maximum?
> > 
> > 
> > I don't really know. I think so.
> 
> It is 50 for the subject and 72 for the body, I changed it accordingly

I just checked. No, 50 is wrong. It is 72 for both the subject and body. At least on GitHub.
But this is also what you find similarly for the Linux kernel, which you can kind of see as the gold standard.
[Here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/submitting-patches.rst?id=bc7938deaca7f474918c41a0372a410049bd4e13#n664):
> For these reasons, the ``summary`` must be no more than 70-75 characters

**Comment by JackTemaki** [comment]
> Ah okay, I simply write my commit messages into the terminal. There are no such assisting tools.

you can add:
```
[core]
    editor = vim
```
in your `.gitconfig` file in your home folder. If you then call `git commit` it will open vim with a pre-defined template and correct colouring.

**Comment by albertz** [comment]
> Ah okay, I simply write my commit messages into the terminal. There are no such assisting tools.

Never do that (i.e. never use `-m`). Always use some editor.

**Comment by JackTemaki** [comment]
> > > > Is 72 the correct maximum?
> > > 
> > > 
> > > I don't really know. I think so.
> > 
> > 
> > It is 50 for the subject and 72 for the body, I changed it accordingly
> 
> I just checked. No, 50 is wrong. It is 72 for both the subject and body. At least on GitHub.
> But this is also what you find similarly for the Linux kernel, which you can kind of see as the gold standard.
> [Here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/submitting-patches.rst?id=bc7938deaca7f474918c41a0372a410049bd4e13#n664):
> 
> > For these reasons, the `summary` must be no more than 70-75 characters

But Github does give a warning when you specify summaries greater than 50 characters, somebody posted a screenshot of this here: https://github.com/microsoft/vscode/issues/60300

**Comment by albertz** [comment]
> > Ah okay, I simply write my commit messages into the terminal. There are no such assisting tools.
> 
> you can add:
> 
> ```
> [core]
>     editor = vim
> ```
> 
> in your `.gitconfig` file in your home folder. If you then call `git commit` it will open vim with a pre-defined template and correct colouring.

This is the default already, or not? Or rather, it will check for `$EDITOR`, which is even better. You better should not overwrite this in the `.gitconfig`. E.g. some environments will overwrite `$EDITOR`, and then Git should follow this, and not be overwritten.

**Comment by albertz** [comment]
> > I just checked. No, 50 is wrong. It is 72 for both the subject and body. At least on GitHub.
> > But this is also what you find similarly for the Linux kernel, which you can kind of see as the gold standard.
> > [Here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/submitting-patches.rst?id=bc7938deaca7f474918c41a0372a410049bd4e13#n664):
> > > For these reasons, the `summary` must be no more than 70-75 characters
> 
> But Github does give a warning when you specify summaries greater than 50 characters, somebody posted a screenshot of this here: [microsoft/vscode#60300](https://github.com/microsoft/vscode/issues/60300)
> 
> But we can go with 72, I also see this as more convenient.

It's not really a warning. It is a "pro tip". This is also the same recommendation by Linus and for Linux.

> So shoot for 50 characters, but consider 72 the hard limit.

This is what I would also recommend. It doesn't necessarily have to be exactly 50 though.

**Comment by albertz** [comment]
Merging now. @npwynands Next time, please make the Git message consistent to our normal style.

**Comment by albertz** [comment]
@npwynands Note that the mail you use for the commit (RWTH) is not registered in your GitHub account. You might want to add that if you want that this commit can be associated with your account.

**Pull request merged**